### PR TITLE
[core] add json kwarg to http request

### LINKED ISSUE #16656: HttpRequest constructor should accept a "json" kwargs
Which is mostly syntactic sugar for `set_json_body`.

This issue implies adding "Content-Type: application/json" to `set_json_body`.

Beware of headers + json as kwargs.

### PR DESCRIPTION
**Pull request by iscai-msft**
fixes #16656

### CODE REVIEW

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py [hunk 1]
```diff
@@ -220,18 +220,32 @@ class HttpRequest(object):
...
+    def __init__(self, method, url, headers=None, files=None, data=None, **kwargs):
+        # type: (str, str, Mapping[str, str], Any, Any, Any) -> None
         self.method = method
         self.url = url
         self.headers = _case_insensitive_dict(headers)
         self.files = files
-        self.data = data
         self.multipart_mixed_info = None  # type: Optional[Tuple]
+        json_kwarg = kwargs.pop("json", None)
+        if data and json_kwarg:
```
Comment (lmazuel) [inline sdk/core/azure-core/azure/core/pipeline/transport/_base.py:239]: Let's say a ValueError

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py [hunk 2]
```diff
@@ -220,18 +220,32 @@ class HttpRequest(object):
...
+            _LOGGER.warning(
+                "You have sent data through both parameter data and kwarg json. "
+                "We will use the data you sent through the data parameter, but "
+                "still conduct json serialization on your inputted data. "
+                "In the future, just use one of these parameters."
+            )
+        self.data = data or json_kwarg
```
Comment (lmazuel) [inline sdk/core/azure-core/azure/core/pipeline/transport/_base.py:246]: That sounds weird to do that. I would have done:
```python
if json_kwarg:
    self.set_json_body(json_kwarg)
else:
    self.data = data
```

File: sdk/core/azure-core/tests/test_pipeline.py [hunk 3]
```diff
@@ -234,14 +234,38 @@ def test_format_incorrect_endpoint(self):
...
+        request.set_json_body(data)
+
+        self.assertEqual(request.data, json.dumps(data))
+        self.assertEqual(request.headers.get("Content-Length"), "17")
+        self.assertEqual(request.headers.get("Content-Type"), "shouldBeMe")
+
+    def test_request_json_kwarg_no_content_type_header(self):
+        request = HttpRequest("GET", "/", json={"documents": ["hello"]})
+
+        assert json.loads(request.data)
```
Comment (lmazuel) [inline sdk/core/azure-core/tests/test_pipeline.py:259]: You should check this is the initial data too

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py (hunk 2 above)
Comment (iscai-msft) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py:246]: oh good point, didn't read that code clearly enough. thanks!

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py [hunk 4]
```diff
@@ -361,13 +375,19 @@ def set_xml_body(self, data):
     def set_json_body(self, data):
         """Set a JSON-friendly object as the body of the request.
 
+        Sets your content type header to "application/json" if you
+        didn't input any content type headers
+
         :param data: A JSON serializable object
+        :raises: TypeError
```
Comment (lmazuel) [inline sdk/core/azure-core/azure/core/pipeline/transport/_base.py:382]: In what circumstances?

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py [hunk 5]
```diff
@@ -220,18 +220,32 @@ class HttpRequest(object):
     :param str url: At least complete scheme/host/path
     :param dict[str,str] headers: HTTP headers
     :param files: Files list.
-    :param data: Body to be sent.
+    :param data: Body to be sent. If you want to send a json body, you should use the
+     `json` kwarg instead. We will handle json serialization for your data.
     :type data: bytes or str.
+    :keyword json: A JSON serializable object. Serializes your inputted object. Use this
+     instead of data if you wish for us to handle json serialization of your object for you.
+    :raises: TypeError
```
Comment (lmazuel) [inline sdk/core/azure-core/azure/core/pipeline/transport/_base.py:228]: Precise the circumstances in which you can get that

File: sdk/core/azure-core/CLIENT_LIBRARY_DEVELOPER.md [hunk 6]
```diff
@@ -193,7 +193,7 @@ The HttpRequest has the following API. It does not vary between transports:
 ```python
 class HttpRequest(object):
 
-    def __init__(self, method, url, headers=None, files=None, data=None):
+    def __init__(self, method, url, headers=None, files=None, data=None, **kwargs):
```
Comment (xiangyan99) [inline sdk/core/azure-core/CLIENT_LIBRARY_DEVELOPER.md:196]: Do we want to accept arbitrary arguments here?

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py [hunk 7]
```diff
@@ -361,13 +375,19 @@ def set_xml_body(self, data):
...
         :param data: A JSON serializable object
+        :raises: TypeError
         """
         if data is None:
             self.data = None
         else:
             self.data = json.dumps(data)
             self.headers["Content-Length"] = str(len(self.data))
+        if not self.headers.get("content-type"):
+            self.headers["Content-Type"] = "application/json"
```
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/pipeline/transport/_base.py:390]: "content-type"? "Content-Type"?

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py (hunk 1 above)
Comment (johanste) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py:239]: What is `json_kwarg` is `{}`? Or False? Or data is 0? If we want to check (and raise) if the user specified these values, we can either check if the values are `in kwargs` or use a different sentinel value than None/compare fals:y values. 

Interestingly enough, I'm not sure if requests raises an exception if you provide both `data` and `json`....

Comment (johanste) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py:239]: (and technically, `None`/`null` is valid json - so there is a difference between not providing the json keyword argument and providing it with a `None` value, so we should distinguish if any value (including `None) was provided or not regardless of if we want to throw or not)

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py (hunk 5 above)
Comment (johanste) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py:228]: I assume that `files` is yet another mutually exclusive argument, correct?

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py:228]: > I assume that `files` is yet another mutually exclusive argument, correct?

This is a good question.

Now if user has 

```py
request = HttpRequest(method, url, files=xxx)
request.set_json_body(json)
```

It is valid but we will set files to None.

If we treat 
```py
request = HttpRequest(method, url, files=xxx, json=yyy)
```
as
```py
request = HttpRequest(method, url, files=xxx)
request.set_json_body(json)
```
Is it the right behavior to override `files` silently?

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py (hunk 4 above)
Comment (iscai-msft) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py:382]: Raises one when you try to json.dumps a non serializable object, i.e. a bytes string (for non 2.7)

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py (hunk 7 above)
Comment (iscai-msft) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py:390]: headers is a case insensitive dict, so searching with case doesn't matter. but you're right, it does look a little weird to just randomly switch casing from line to line, so will make 389 capitalized as well, thanks!

File: sdk/core/azure-core/CLIENT_LIBRARY_DEVELOPER.md (hunk 6 above)
Comment (iscai-msft) [reply sdk/core/azure-core/CLIENT_LIBRARY_DEVELOPER.md:196]: we're adding kwargs to the `HttpRequest` so I thought I would also add it to client library developer

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py (hunk 1 above)
Comment (iscai-msft) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py:239]: @johanste yeah, just checked `requests` behavior, and it allow us to pass in both, but it actually seems to ignore `json` when both are passed

image

i'm not sure how closely we're trying to model `HttpRequest` after `requests`, but I think it'd be fair to follow `requests` here

Comment (iscai-msft) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py:239]: decided offline with @johanste and @lmazuel that we will raise a `ValueError`

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py (hunk 5 above)
Comment (iscai-msft) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py]: @xiangyan99 following the fact that we will raise if users pass in `data` and `json`, we should raise if `files` and `json` are specified

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py [hunk 8]
```diff
@@ -220,18 +220,31 @@ class HttpRequest(object):
...
+        # type: (str, str, Mapping[str, str], Any, Any, Any) -> None
         self.method = method
         self.url = url
         self.headers = _case_insensitive_dict(headers)
         self.files = files
-        self.data = data
         self.multipart_mixed_info = None  # type: Optional[Tuple]
+        json_kwarg = "json" in kwargs
+        body_input = [files, data]
+        if json_kwarg and len([b for b in body_input if b]) > 0:
```
Comment (iscai-msft) [inline sdk/core/azure-core/azure/core/pipeline/transport/_base.py:240]: Ideally I would throw a `ValueError` if more than one of `files`, `data`, or `json` is specified. However, I fear throwing if `files` and `data` are specified would be a breaking change, cc @johanste for confirmation

File: sdk/core/azure-core/CHANGELOG.md [hunk 9]
```diff
@@ -1,7 +1,10 @@
 # Release History
 
-## 1.11.1 (Unreleased)
+## 1.12.0 (Unreleased)
 
+### Features
+
+- Add `json` kwarg to `HttpRequest` object. The `json` kwarg takes your data as input and json serializes it for you. Use instead of `data` if you want this behavior. #16695
```
Comment (lmazuel) [inline sdk/core/azure-core/CHANGELOG.md:7]: Is there a missing word, or do I don't understand the sentence?

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py (hunk 8 above)
Comment (lmazuel) [inline sdk/core/azure-core/azure/core/pipeline/transport/_base.py:240]:
```python
if json_kwarg and any(body_input):
```

Comment (iscai-msft) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py]: lmao def better, thanks!

File: sdk/core/azure-core/CHANGELOG.md (hunk 9 above)
Comment (iscai-msft) [reply sdk/core/azure-core/CHANGELOG.md]: updated the changelog wording, does the new wording sound better to you?

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py (hunk 5 above)
Comment (johanste) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py]: `files` imply multipart, correct?

Comment (johanste) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py]: If so, yeah, while possible that one or more parts are json, I would still expect that to come in through `files`.

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py (hunk 8 above)
Comment (johanste) [reply sdk/core/azure-core/azure/core/pipeline/transport/_base.py]: Let's start with a `warnings.warn`.

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py [hunk 10]
```diff
@@ -220,18 +221,36 @@ class HttpRequest(object):
...
     :type data: bytes or str.
+    :keyword json: A JSON serializable object. Serializes your inputted object. Use this
+     instead of data if you wish for us to handle json serialization of your object for you.
+    :raises: TypeError if you input a non json serializable object through kwarg `json`
     """
 
-    def __init__(self, method, url, headers=None, files=None, data=None):
-        # type: (str, str, Mapping[str, str], Any, Any) -> None
+    def __init__(self, method, url, headers=None, files=None, data=None, **kwargs):
+        # type: (str, str, Mapping[str, str], Any, Any, Any) -> None
```
Comment (johanste) [inline sdk/core/azure-core/azure/core/pipeline/transport/_base.py:233]: `**Any` for `**kwargs`.

File: sdk/core/azure-core/azure/core/pipeline/transport/_base.py [hunk 11]
```diff
@@ -220,18 +221,36 @@ class HttpRequest(object):
...
+                "in a value of 'files' or 'data'."
+            )
+        if files and data:
+            warnings.warn(
+                "You have specified values for both data and files. "
+                "Please only put in values for data or for files."
+            )
+        self.data = data or kwargs.pop("json", None)
+        if json_kwarg:
+            self.set_json_body(self.data)
```
Comment (johanste) [inline sdk/core/azure-core/azure/core/pipeline/transport/_base.py:253]: Let's raise if there are **kwargs that we haven't consumed.

**Comment by iscai-msft** [comment]
putting on the backburner, closing this PR

**Pull request closed without merging**